base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1460 @@
+         arrayset_typecheck(aryt, valt))
+        add_thrown_escapes!(astate, pc, args, 2)
+    end
+    ary = args[3]
+    val = args[4]
+    inbounds = isa(boundcheckt, Const) && !boundcheckt.val::Bool
+    inbounds || add_escape_change!(astate, ary, ThrownEscape(pc))
+    # we don't track precise index information about this array and won't record what value
+    # is being assigned here: directly propagate the escape information of this array to
+    # the value being assigned as the most conservative propagation
+    # TODO enable index analysis when constant values are available?
+    estate = astate.estate
+    if isa(ary, SSAValue) || isa(ary, Argument)
+        aryinfo = estate[ary]
+    else
+        # unanalyzable object (e.g. obj::GlobalRef): escape field value conservatively
+        add_escape_change!(astate, val, ⊤)
+        return true
+    end
+    AliasEscapes = aryinfo.AliasEscapes
+    if isa(AliasEscapes, Bool)
+        if !AliasEscapes
+            # the elements of this array haven't been analyzed yet: don't need to consider ArrayEscapes for now
+            @goto add_ary_escape
+        end
+        @label conservative_propagation
+        add_escape_change!(astate, val, aryinfo)
+    elseif isa(AliasEscapes, ArrayEscapes)
+        for xidx in AliasEscapes
+            if isa(xidx, Int)
+                x = irval(xidx, estate)::SSAValue # TODO remove me once we implement ArgEscape
+                add_escape_change!(astate, val, estate[x])
+                add_alias_change!(astate, val, x)
+            else
+                add_escape_change!(astate, val, ThrownEscape((xidx::SSAValue).id))
+            end
+        end
+        @label add_ary_escape
+        add_escape_change!(astate, val, ignore_aliasescapes(aryinfo))
+    else
+        # this object has been used as struct, but it is "used" as array here (thus should throw)
+        # update ary's element information and just handle this case conservatively
+        @assert isa(AliasEscapes, FieldEscapes)
+        aryinfo = escape_unanalyzable_obj!(astate, ary, aryinfo)
+        @goto conservative_propagation
+    end
+    # also propagate escape information imposed on the return value of this `arrayset`
+    ssainfo = estate[SSAValue(pc)]
+    add_escape_change!(astate, ary, ssainfo)
+    return true
+end
+
+function escape_builtin!(::typeof(arraysize), astate::AnalysisState, pc::Int, args::Vector{Any})
+    length(args) == 3 || return false
+    ary = args[2]
+    dim = args[3]
+    if !arraysize_typecheck(ary, dim, astate.ir)
+        add_escape_change!(astate, ary, ThrownEscape(pc))
+        add_escape_change!(astate, dim, ThrownEscape(pc))
+    end
+    # NOTE we may still see "arraysize: dimension out of range", but it doesn't capture anything
+    return true
+end
+
+function arraysize_typecheck(@nospecialize(ary), @nospecialize(dim), ir::IRCode)
+    aryt = argextype(ary, ir)
+    aryt ⊑ₜ Array || return false
```

Comment:
This symbol didn't render in my browser

Reply:
In the module context of `EscapeAnalysis` there are
- `a \sqsubseteq b`: escape lattice partial order
- `a \sqsubseteq_{t} b`: type lattice partial order

Maybe we want to use different symbols here?

Reply:
It seems unfortunate to change the symbol for the regular type partial order, maybe invert it and use `⊑ₑ` for the escape lattice?

Reply:
Yea, for now it seems better to rename escape lattice operations.
Once we finish our type lattice overhaul, I think `x::LatticeElm ⊑ y::Lattice` and `x::EscapeInfo ⊑ y::EscapeInfo` can safely share the same symbol without loosing any performance though.

Reply:
9f09787a6b0ad42ac441dba6228d7772e8889e59